File: examples/primitives/address/checksum-validation.ts
/**
 * EIP-55 Checksum Validation Example
 *
 * Demonstrates:
 * - EIP-55 checksummed address generation
 * - Checksum validation and verification
 * - Detecting invalid checksums to prevent typos
 * - Safe parsing with checksum requirements
 */

import { Address } from "../../../src/primitives/Address/index.js";

const addr = new Address("0x742d35cc6634c0532925a3b844bc9e7595f51e3e");

const validChecksummed = "0x742d35Cc6634C0532925a3b844Bc9e7595f51e3e";
const invalidChecksum = "0x742d35cc6634C0532925a3b844Bc9e7595f51e3e"; // Wrong case
const allLowercase = "0x742d35cc6634c0532925a3b844bc9e7595f51e3e";
const allUppercase = "0x742D35CC6634C0532925A3B844BC9E7595F51E3E";

function safeParseAddress(input: string): Address | null {
	// Validate format first
	if (!Address.isValid(input)) {
		return null;
	}

	// If mixed case, validate checksum
	const hasMixedCase =
		input !== input.toLowerCase() && input !== input.toUpperCase();
	if (hasMixedCase && !Address.isValidChecksum(input)) {
		return null;
	}

	const addr = Address.fromHex(input);
	return addr;
}

// Valid addresses
safeParseAddress("0x742d35Cc6634C0532925a3b844Bc9e7595f51e3e"); // Valid checksum
safeParseAddress("0x742d35cc6634c0532925a3b844bc9e7595f51e3e"); // Lowercase (no checksum)

// Typo - one wrong character case
safeParseAddress("0x742d35Cc6634C0532925a3b844Bc9e7595f51E3e"); // Last 'E' should be lowercase

function requireChecksum(input: string): Address {
	if (!Address.isValidChecksum(input)) {
		throw new Error(`Address must have valid EIP-55 checksum: ${input}`);
	}

	// Ensure it's actually checksummed (not all same case)
	const hasMixedCase =
		input !== input.toLowerCase() && input !== input.toUpperCase();
	if (!hasMixedCase) {
		throw new Error(`Address must be checksummed (mixed case): ${input}`);
	}

	return Address.fromHex(input);
}

try {
	const strictAddr = requireChecksum(
		"0x742d35Cc6634C0532925a3b844Bc9e7595f51e3e",
	);
} catch (e) {}

try {
	requireChecksum("0x742d35cc6634c0532925a3b844bc9e7595f51e3e"); // All lowercase
} catch (e) {}

const addresses = [
	"0x742d35cc6634c0532925a3b844bc9e7595f51e3e",
	"0xA0CF798816D4B9B9866B5330EEA46A18382F251E",
	"0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266",
];
for (const addrStr of addresses) {
	const addr = Address.fromHex(addrStr);
	const checksummed = addr.toChecksummed();
}

function validateUserInput(input: string): {
	valid: boolean;
	address?: Address;
	error?: string;
} {
	// Trim whitespace
	const trimmed = input.trim();

	// Check basic format
	if (!Address.isValid(trimmed)) {
		return { valid: false, error: "Invalid address format" };
	}

	// Check checksum if mixed case
	const hasMixedCase =
		trimmed !== trimmed.toLowerCase() && trimmed !== trimmed.toUpperCase();
	if (hasMixedCase && !Address.isValidChecksum(trimmed)) {
		// Provide the correct checksummed version
		const addr = Address.fromHex(trimmed);
		const correct = addr.toChecksummed();
		return {
			valid: false,
			error: `Invalid checksum. Did you mean: ${correct}?`,
		};
	}

	const address = Address.fromHex(trimmed);
	return { valid: true, address };
}

const userInputs = [
	"0x742d35Cc6634C0532925a3b844Bc9e7595f51e3e", // Correct
	"0x742d35cc6634C0532925a3b844Bc9e7595f51e3e", // Wrong checksum
	"  0x742d35cc6634c0532925a3b844bc9e7595f51e3e  ", // With whitespace
	"0xINVALID",
];

for (const input of userInputs) {
	const result = validateUserInput(input);
	if (result.valid) {
	} else {
	}
}

File: packages/voltaire-zig/examples/primitives/address/checksum-validation.zig
const std = @import("std");
const primitives = @import("primitives");
const Address = primitives.address.Address;

pub fn main() !void {
    std.debug.print("\n=== EIP-55 Checksum Validation ===\n\n", .{});

    // 1. Understanding EIP-55 checksums
    std.debug.print("1. EIP-55 Checksumming\n\n", .{});

    const addr = try Address.fromHex("0x742d35cc6634c0532925a3b844bc9e7595f51e3e");

    std.debug.print("EIP-55 uses mixed-case encoding for error detection:\n", .{});
    const lowercase = Address.toHex(addr);
    const checksummed = Address.toChecksummed(addr);
    std.debug.print("Original (lowercase): {s}\n", .{&lowercase});
    std.debug.print("Checksummed:         {s}\n", .{&checksummed});
    std.debug.print("Notice: Some letters capitalized based on keccak256 hash\n\n", .{});

    // 2. Checksum validation
    std.debug.print("2. Checksum Validation\n\n", .{});

    const valid_checksummed = "0x742d35Cc6634C0532925a3b844Bc9e7595f51e3e";
    const invalid_checksum = "0x742d35cc6634C0532925a3b844Bc9e7595f51e3e"; // Wrong case
    const all_lowercase = "0x742d35cc6634c0532925a3b844bc9e7595f51e3e";
    const all_uppercase = "0x742D35CC6634C0532925A3B844BC9E7595F51E3E";

    std.debug.print("Valid checksummed:   {}\n", .{Address.isValidChecksum(valid_checksummed)});
    std.debug.print("Invalid checksum:    {}\n", .{Address.isValidChecksum(invalid_checksum)});
    std.debug.print("All lowercase:       {}\n", .{Address.isValidChecksum(all_lowercase)}); // true - no checksum
    std.debug.print("All uppercase:       {}\n\n", .{Address.isValidChecksum(all_uppercase)}); // true - no checksum

    // 3. Detecting typos with checksums
    std.debug.print("3. Typo Detection\n\n", .{});

    // Helper to check if string has mixed case
    const has_mixed_case = struct {
        fn check(s: []const u8) bool {
            var has_upper = false;
            var has_lower = false;
            for (s) |c| {
                if (c >= 'A' and c <= 'F') has_upper = true;
                if (c >= 'a' and c <= 'f') has_lower = true;
            }
            return has_upper and has_lower;
        }
    }.check;

    // Valid addresses
    const test_addrs = [_][]const u8{
        "0x742d35Cc6634C0532925a3b844Bc9e7595f51e3e", // Valid checksum
        "0x742d35cc6634c0532925a3b844bc9e7595f51e3e", // Lowercase (no checksum)
        "0x742d35Cc6634C0532925a3b844Bc9e7595f51E3e", // Typo - last 'E' wrong
    };

    for (test_addrs) |test_addr| {
        if (!Address.isValid(test_addr)) {
            std.debug.print("✗ Invalid format: {s}\n", .{test_addr});
            continue;
        }

        // If mixed case, validate checksum
        const mixed = has_mixed_case(test_addr);
        if (mixed and !Address.isValidChecksum(test_addr)) {
            std.debug.print("✗ Invalid checksum (possible typo): {s}\n", .{test_addr});
            continue;
        }

        const parsed = try Address.fromHex(test_addr);
        const parsed_checksum = Address.toChecksummed(parsed);
        std.debug.print("✓ Valid address: {s}\n", .{&parsed_checksum});
    }
    std.debug.print("\n", .{});

    // 4. Converting addresses to checksummed format
    std.debug.print("4. Converting to Checksummed Format\n\n", .{});

    const addresses = [_][]const u8{
        "0x742d35cc6634c0532925a3b844bc9e7595f51e3e",
        "0xA0CF798816D4B9B9866B5330EEA46A18382F251E",
        "0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266",
    };

    std.debug.print("Converting addresses to EIP-55 checksummed format:\n", .{});
    for (addresses) |addr_str| {
        const parsed = try Address.fromHex(addr_str);
        const checksummed_out = Address.toChecksummed(parsed);
        std.debug.print("{s}\n", .{addr_str});
        std.debug.print("→ {s}\n\n", .{&checksummed_out});
    }

    // 5. Comparison: case-insensitive equality
    std.debug.print("5. Case-Insensitive Equality\n\n", .{});

    const addr1 = try Address.fromHex("0x742d35Cc6634C0532925a3b844Bc9e7595f51e3e");
    const addr2 = try Address.fromHex("0x742d35cc6634c0532925a3b844bc9e7595f51e3e");
    const addr3 = try Address.fromHex("0x742D35CC6634C0532925A3B844BC9E7595F51E3E");

    std.debug.print("Address 1: {s}\n", .{&Address.toChecksummed(addr1)});
    std.debug.print("Address 2: {s}\n", .{&Address.toHex(addr2)});
    std.debug.print("Address 3: {s}\n", .{&Address.toUppercase(addr3)});
    std.debug.print("All equal (case-insensitive): {}\n", .{
        Address.equals(addr1, addr2) and Address.equals(addr2, addr3),
    });
    std.debug.print("\n", .{});

    // 6. Best practices
    std.debug.print("6. Best Practices\n\n", .{});
    std.debug.print("✓ Always display addresses in checksummed format\n", .{});
    std.debug.print("✓ Validate checksums on user input when mixed-case\n", .{});
    std.debug.print("✓ Store addresses as bytes internally (case-insensitive)\n", .{});
    std.debug.print("✓ Convert to checksummed for display/logging\n", .{});
}
